File: java/scf-bp-component/src/main/java/org/mydotey/scf/bp/component/labeled/DynamicDcConfigurationSource.java
package org.mydotey.scf.bp.component.labeled;

import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

import org.mydotey.scf.ConfigurationSourceConfig;

/**
 * @author koqizhao
 *
 * May 17, 2018
 */
public class DynamicDcConfigurationSource extends AbstractDcConfigurationSource {

    private Map<DcSetting, DcSetting> _dcSettings = new ConcurrentHashMap<>();

    public DynamicDcConfigurationSource(ConfigurationSourceConfig config) {
        super(config);
    }

    public void updateSetting(DcSetting setting) {
        Objects.requireNonNull(setting, "setting is null");
        Objects.requireNonNull(setting.getKey(), "setting.key is null");

        DcSetting oldValue = _dcSettings.get(setting);
        if (oldValue != null) {
            if (Objects.equals(oldValue.getValue(), setting.getValue()))
                return;
        }

        setting = setting.clone();
        _dcSettings.put(setting, setting);

        raiseChangeEvent();
    }

    public void removeSetting(DcSetting setting) {
        Objects.requireNonNull(setting, "setting is null");
        Objects.requireNonNull(setting.getKey(), "setting.key is null");

        DcSetting oldValue = _dcSettings.remove(setting);

        if (oldValue == null)
            return;

        raiseChangeEvent();
    }

    @Override
    protected DcSetting getPropertyValue(DcSetting setting) {
        return _dcSettings.get(setting);
    }

}

File: dotnet/src/MyDotey.SCF.BP.Component/Labeled/DynamicDcConfigurationSource.cs
using System;
using System.Collections.Concurrent;

namespace MyDotey.SCF.BP.Component.Labeled
{
    /**
     * @author koqizhao
     *
     * May 17, 2018
     */
    public class DynamicDcConfigurationSource : AbstractDcConfigurationSource
    {
        private ConcurrentDictionary<DcSetting, DcSetting> _dcSettings = new ConcurrentDictionary<DcSetting, DcSetting>();

        public DynamicDcConfigurationSource(ConfigurationSourceConfig config)
            : base(config)
        {

        }

        public virtual void UpdateSetting(DcSetting setting)
        {
            if (setting == null)
                throw new ArgumentNullException("setting is null");

            if (string.IsNullOrWhiteSpace(setting.Key))
                throw new ArgumentNullException("setting.key is null");

            _dcSettings.TryGetValue(setting, out DcSetting oldValue);
            if (oldValue != null)
            {
                if (object.Equals(oldValue.Value, setting.Value))
                    return;
            }

            setting = (DcSetting)setting.Clone();
            _dcSettings[setting] = setting;

            RaiseChangeEvent();
        }

        public virtual void RemoveSetting(DcSetting setting)
        {
            if (setting == null)
                throw new ArgumentNullException("setting is null");

            if (string.IsNullOrWhiteSpace(setting.Key))
                throw new ArgumentNullException("setting.key is null");

            _dcSettings.TryRemove(setting, out DcSetting oldValue);

            if (oldValue == null)
                return;

            RaiseChangeEvent();
        }

        protected override DcSetting GetPropertyValue(DcSetting setting)
        {
            _dcSettings.TryGetValue(setting, out DcSetting value);
            return value;
        }
    }
}